### ISSUE TITLE
Implement WorldEdit Selection Tool &| Improve Selector.

### ISSUE THREAD

**Issue by SirMangler** [TechFortress/GriefPrevention#561]
**AS** An administrator and player.
**I WANT** The ability to use the WorldEdit wand.
**SO THAT** I can not lose my sanity trying to work the claims system.

I like this plugin a lot, but the selection tool is incredibly difficult to use. I spent 15 minutes trying to claim my Spawn rectangle when it could have been done with WorldEdit in 1 minute. Using WorldEdit's API, you should easily be able to access the area of blocks. 

One of the biggest problems, the selection tool immediately claims the area rather than allowing you to fine-tune, and if you put the tool away, it cancels (including the admin mode). 
Given the ability to select the blocks when I want or need without it being interrupted by a cluck of a chicken would really help. In the end I didn't actually manage to claim my spawn (which is worldguarded) to prevent overlapping claims, simply because I don't know what it's trying to do.

It could be that despite I'm admin, I have no remaining blocks (it didn't say that was a problem though), or that it was only trying to claim a single block rather than both corners. Or it could be something related to the fact it had made a claim without me wanting it to in only one corner and it wasn't expanding in the intended way. Or maybe it was trying to claim based on my movement and not the corner-to-corner system?

I appreciate your response might be "just read the wiki", and I did. All I mean by this, is that it can be made much simpler, especially since this is for regular players.

Thanks for this plugin. I'll definitely reinstall this plugin if the selection tool is improved.

**Comment by FrozenBeard** [comment]
Ive requested this multiple times. Also for making a selection in WE to do mass claim deletions.

**Comment by RoboMWM** [comment]
> I WANT The ability to use the WorldEdit wand

Use it how

**Comment by SirMangler** [comment]
> 
> 
> > I WANT The ability to use the WorldEdit wand
> 
> Use it how

I appreciate you've probably not bothered to read my post at all, perhaps due to the length. I'll put it simply.

```
By Functionality:
World Edit Wand > GriefPrevention Wand
```

**Comment by RoboMWM** [comment]
I was on the phone, so I didn't feel like tapping out the reason, but it's basically that the story format there is to make it easier to see and remember what the feature was about in a glance; i.e., it's not because I didn't read what you put, but I wanted you to edit that there.

**Comment by RoboMWM** [comment]
You bring up a bunch of different issues in your post, it seems more like feedback rather than a specific request.

> the selection tool immediately claims the area rather than allowing you to fine-tune,

You can fine-tune via selecting any corner and clicking where you want to move it.

> it was only trying to claim a single block rather than both corners. 

Sometimes the MC client/mouse sends an interact event twice (i.e. a right click twice), causing it to set both corners at the same time. For admin claims, there is no minimum size limit for a claim so this is allowed.

> Given the ability to select the blocks when I want or need without it being interrupted by a cluck of a chicken would really help. 

What's the issue here?

WorldEdit is a world-editing tool, not too dissimilar from image tools like paint.NET. I sure hope it's better. Also, GP is intended not to interfere with the tool's functionality (although now shovels and even axes have right-click features). Also, GP's claims are 2D, containing only the X and Z coordinates. So you just only need to worry about area, not volume.

**Comment by SirMangler** [comment]
World Edit comes with a ton of [selection specific commands](https://minecraft-worldedit.fandom.com/wiki/Worldedit_Commands#Region_Selection). 

> Sometimes the MC client/mouse sends an interact event twice (i.e. a right click twice)

If this is the case, I've never had this problem before, with WorldEdit, MV-P, or my own plugins for that matter.

> What's the issue here?

The issue is the selection system (I'm talking about corner to corner mostly), can be improved.

I mentioned WorldEdit because it saves the trouble of redoing the selection tool yourselves, and if you or anyone prefers it, they don't have to lose the current system, while pleasing the rest who are used to the WE wand and would much prefer it.

If you want exact suggestions:

- Add Left and Right Click for Pos1 and Pos2, and the appropriate feedback in chat.
- Stop it from cancelling all commands and selections if you put away the tool.
- Add the ability to confirm the selection, rather than simply just making it.

After further testing, I've discovered that "resize claim" moves the corners, rather than creating a new selection box. That's probably where the bulk of the problem came for me. Being used to World Edit, I didn't even think of it. 

I apologise that I wasn't clear in my message and that I was rude and really shouldn't have been, the selection tool made me extremely frustrated on the day I tried it. (it creating multiple claims, claiming the wrong areas, then not being able to find where it actually claimed, so I had to delete all claims and start again). The problem was I needed to claim top-to-bottom, again a selection that could be made in world edit in about 10 seconds.

**Comment by Jikoo** [comment]
>>Sometimes the MC client/mouse sends an interact event twice (i.e. a right click twice)

>If this is the case, I've never had this problem before, with WorldEdit, MV-P, or my own plugins for that matter.

I'm at a loss for what's causing double registration in GP, GP is properly checking which hand is interacting and what's in it. I've definitely noticed it happen sometimes when users try to set a claim corner on a block that's within reach. I always tell them to just back up to the point where their client can't reach the block, but it doesn't happen with other plugins.

> * Add Left and Right Click for Pos1 and Pos2, and the appropriate feedback in chat.

That would eliminate the possibility of double click registration issues, I'm a fan 👍 

> * Stop it from cancelling all commands and selections if you put away the tool.

Putting away the tool losing claim state is very frustrating, I'll admit, but one of GP's selling points is the "no commands required" system where users just point and click. How would you suggest resetting claim state without commands? It'd be nice to have a grace period with a warning, but I don't want to see the system vanish entirely.
For a workaround, you can swap the shovel into the off hand without interrupting claiming last I was aware. It's not great, but it makes it a little easier to avoid accidentally finishing a claim.

I do agree that the aggressiveness of swapping off admin claim mode is annoying, but I believe that GP's stance on admin abilities not being eternal toggles is correct. The amount of accidental admin ability abuse (opening locked chests, etc.) dropped drastically when the server I used to manage made the change to GP.

> * Add the ability to confirm the selection, rather than simply just making it.

How would you propose that that be done without commands? I don't dislike the idea, but if double registers are already an issue for clicks, the confirmation can't just be "click again to confirm selection" or something of that nature. I suppose a scheduled wait before confirmation could work around that.

**Comment by RoboMWM** [comment]
When I asked "what's the issue here?" I specifically meant what I quoted:

> Given the ability to select the blocks when I want or need without it being interrupted by a cluck of a chicken would really help.

The sentence seems incomplete and I'm not sure how the cluck of a chicken interrupts anything.

Minor point aside, Jikoo raises good questions. 

In regards to WorldEdit having selection commands - that illustrates my point about WorldEdit providing tools to modify the world exactly as you see fit, as many of its commands operate within a selection. GP, on the contrary, only cares about the 2D area and is meant to be simple. I did advocate for and eventually got the `/claim` and `/expand` commands as I mentioned it's near-impossible to perform point-and-click methods when underground, as the claim outline blocks currently are suited to appear at the surface. (There is also another issue pointing out that they don't work too well for islands/skyblock modes either.)

> I'm at a loss for what's causing double registration in GP, GP is properly checking which hand is interacting and what's in it.

This can happen if the client sees a block that the server doesn't. I.e., in the snow, the outline blocks are taller than the snow blocks. When a player is pointing at the upper part of the block, the server may believe they're actually pointing at something else. When an interact event occurs, there will be one for what the client sees, and then another one will be fired for what is really there on the server. (It may also be the case that the server sends a block update, and then the client resends the interact - but iirc this seems to happen too quickly for that to be the case.)
